test/functional/taef/tests.cpp

```diff
@@ -40,6 +42,327 @@
+                    return false;
+                }
+            }
+            PathEnd = strrchr(DriverFilePath, '\\');
+            if (PathEnd == NULL) {
+                TraceError(
+                    "[ lib] ERROR, %s.",
+                    "Failed parsing unexpected module path format");
+                return false;
+            }
+            PathEnd++;
+            size_t RemainingLength = sizeof(DriverFilePath) - (PathEnd - DriverFilePath);
+            int PathResult =
+                snprintf(
+                    PathEnd,
+                    RemainingLength,
+                    "%s.sys",
+                    DriverName);
+            if (PathResult <= 0 || (size_t)PathResult > RemainingLength) {
+                TraceError(
+                    "[ lib] ERROR, %s.",
+                    "Failed to create driver on disk file path");
+                return false;
+            }
+            if (GetFileAttributesA(DriverFilePath) == INVALID_FILE_ATTRIBUTES) {
+                TraceError(
+                    "[ lib] ERROR, %s.",
+                    "Failed to find driver on disk");
+                return false;
+            }
+            ServiceHandle =
+                CreateServiceA(
+                    ScmHandle,
+                    DriverName,
+                    DriverName,
+                    SC_MANAGER_ALL_ACCESS,
+                    SERVICE_KERNEL_DRIVER,
+                    SERVICE_DEMAND_START,
+                    SERVICE_ERROR_NORMAL,
+                    DriverFilePath,
+                    nullptr,
+                    nullptr,
+                    DependentFileNames,
+                    nullptr,
+                    nullptr);
+            if (ServiceHandle == nullptr) {
+                Error = GetLastError();
+                if (Error == ERROR_SERVICE_EXISTS) {
+                    goto QueryService;
+                }
+                TraceError(
+                    "[ lib] ERROR, %u, %s.",
+                    Error,
+                    "CreateService failed");
+                return false;
+            }
+        }
+        return true;
+    }
+    void Uninitialize() {
+        if (ServiceHandle != nullptr) {
+            CloseServiceHandle(ServiceHandle);
+        }
+        if (ScmHandle != nullptr) {
+            CloseServiceHandle(ScmHandle);
+        }
+    }
+    bool Start() {
+        if (!StartServiceA(ServiceHandle, 0, nullptr)) {
+            uint32_t Error = GetLastError();
+            if (Error != ERROR_SERVICE_ALREADY_RUNNING) {
+                TraceError(
+                    "[ lib] ERROR, %u, %s.",
+                    Error,
+                    "StartService failed");
+                return false;
+            }
+        }
+        return true;
+    }
+};
+
+class DriverClient {
+    HANDLE DeviceHandle;
+public:
+    DriverClient() : DeviceHandle(INVALID_HANDLE_VALUE) { }
+    ~DriverClient() { Uninitialize(); }
```

Comment:
we should not use destructors - use WIL resource types to close resources instead.